### DISCUSSION TITLE
Media-Library-Plugin - do not replace existing media in collection

### DISCUSSION THREAD

**Discussion by diogogpinto** [filamentphp/filament#8674] [Ideas]
Hello all 👋🏻

If I'm correct, the [spatie-laravel-media-library-plugin](https://github.com/filamentphp/spatie-laravel-media-library-plugin) allows you to upload a file to a collection. But, if you want to add more files in the future, it currently deletes the files in existence.

Imagine the following workflow:
- User has many documents (contracts, insurance policies, whatever)
- You add one file today
- When, tomorrow, you upload another file with this plugin, it deletes the file you uploaded today

This is because of the saveRelationshipsUsing() method, that calls $component->deleteAbandonedFiles().

What I'm about to do is create a custom field, that extends this one, but with a property -> shouldDeleteOldFiles that when set to false doesn't call $component->deleteAbandonedFiles().

I don't know if this has many requests, but it's a simple feature request that I think can improve this plugin a lot.

Thank you!

**Comment by 8888post** [comment]
```
use Filament\Forms\Components\FileUpload;
 
FileUpload::make('attachments')
    ->multiple()
```
https://filamentphp.com/docs/3.x/forms/fields/file-upload#uploading-multiple-files

**Reply by diogogpinto** [reply]
That has nothing to do with the posted issue. I suggested the ability to add media to a model without replacing the old media attached. Here's my approach:

```php
class ExtendSpatieMediaLibraryFileUpload extends SpatieMediaLibraryFileUpload
{

    protected bool | Closure $shouldReplaceOldFiles = false;

    protected function setUp(): void
    {
        parent::setUp();
    }

    public function replaceOldFiles(bool | Closure $condition = true): static
    {
        $this->shouldReplaceOldFiles = $condition;

        return $this;
    }

    public function deleteAbandonedFiles(): void
    {

        if (!$this->shouldReplaceOldFiles) return;

        /** @var Model&HasMedia $record */
        $record = $this->getRecord();

        $record
            ->getMedia($this->getCollection())
            ->whereNotIn('uuid', array_keys($this->getState() ?? []))
            ->each(fn (Media $media) => $media->delete());
    }
}
```

If you use this field it won't delete the old media by default.

**Reply by OwKe** [reply]
You could also filter the existing media collection, so they are not included inside `deleteAbandonedFiles()`

```
->filterMediaUsing(function (Collection $media): Collection {
    return $media->where('collection_name', '!=', '...');
})
```

**Reply by RomainMazB** [reply]
I totally agree with the feature suggestion.

That said, I confirm the @OwKe trick is a perfect workaround!

Returning an empty Eloquent Collection inside the `filterMediaUsing` to avoid any accidental bad filters and uselessly looping on a collection
```php
use Illuminate\Database\Eloquent\Collection;

SpatieMediaLibraryFileUpload::make('attachments')
  ->collection('attachments')
  ->multiple()
  ->filterMediaUsing(fn (): Collection => new Collection)
```
